File: oneapp/onespace/finding.py
"""One box over every space — what somebody types when they know the name.

The gap this closes is the one `docs/FRAPPE.md` named first: a person who knows
a customer's name and not which space it lives in had nowhere to type it, and
every space was searched by opening it.

**Not Frappe's global search**, and that was a measurement rather than a
preference. `__global_search` on the dev site holds 1,348 rows, of which 1,022
are `DocType`, 223 are `Report` and 55 are `Module Def` — it is the desk's own
metadata, and the business records a person actually looks for are barely in it.
Two things would still be wrong if it were full. It is a MySQL full-text index
queried with `MATCH … AGAINST` in natural-language mode, so it matches whole
words: typing `Meri` does not find `Meridian`, which is fatal for a box people
type four characters into. And a hit is a doctype and an id with no idea which
*screen* shows it — and a record nobody can open is not a search result here,
because `_refuse_ungranted` means a record is only ever reachable through a
screen that grants it.

**So the screens are what is searched.** For each screen this reader could open,
one `like` over the fields that screen lists, and a hit comes back already
knowing where to open it. That is the thing we have and Frappe does not: the
manifest already says which space and which screen a doctype belongs to.

Measured before it was written, because a fan-out over a hundred-odd tables is
the kind of design that sounds slow: one `get_list` with a `like` is about half
a millisecond, and the whole fan-out across all 146 doctypes the shipped
manifests name is **68 ms** for a real query. It is the pathological query —
two characters that match everything — that costs, and that is bounded by
`MOST` per screen and `TOTAL` overall rather than by the number of tables.

Three rules the fan-out inherits rather than re-decides, and all three are the
list's own:

  * **A hit is a row that screen would list.** The screen's declared filters
    apply, `@me` resolved the same way `mine.py` resolves it everywhere else,
    so a search cannot become a way around a narrowing — "my leave" stays mine.
  * **Only fields the screen shows are searched.** `filters.py` says why:
    watching which rows come back is a way of reading a column you were never
    given.
  * **`get_list`, so permissions are Frappe's own.** Nothing here passes
    `ignore_permissions`, and the targets come from `navigable`, which is
    already narrowed to the seat this person holds.

What is deliberately not here: the *other* half of a command palette. Going to
a screen needs no server at all — `api.session` already hands the browser every
space with its screens — and doing anything is `spaceview/actions.py`, which
already knows what a record offers. This module answers one question.
"""

import frappe
from frappe.utils import strip_html_tags

from oneapp.onespace import fieldtypes, mine
from oneapp.onespace.spaceview.filters import (
	MAX_SEARCH_COLUMNS,
	NEVER_SEARCHED,
	_search_text,
)

#: Rows taken from any one screen. A palette is a list somebody reads, so the
#: interesting answer is "which screens have this" rather than "here are forty
#: invoices" — and a screen that floods the answer is a screen that hides the
#: other nine.
MOST = 5

#: And in the whole answer, however many screens matched.
TOTAL = 40

#: Nothing is searched for less than this. Not a performance guard — `zz` costs
#: 490 ms because it matches everything, which is the shape of the cost — but a
#: product one: one letter matches half the workspace and answers nothing.
SHORTEST = 2

#: Screens looked at in one answer. A ceiling rather than a target: the shipped
#: manifests name 146 doctypes between them and a reader holds a fraction of
#: that, so this only ever bites on a workspace with every space enabled and
#: somebody holding every seat in all of them.
SCREENS = 200

#: Where a match landed, best first. A palette is read top-down and the id
#: somebody pasted has to be the first line, not the eleventh.
#:
#: `ELSEWHERE` is the one that is not about the title, and it is the difference
#: between a useful list and a list in arrival order. A row can match on any
#: column the screen shows, so a ledger entry named `09b526cc60` comes back for
#: `merid` because the account it is *against* says so — and with only four
#: ranks it tied with every genuine name match and won on sort order. A hit
#: whose own name does not contain what was typed is a hit somebody cannot see
#: the reason for, so it goes under all of them.
EXACT, STARTS, WORD, ANYWHERE, ELSEWHERE = 0, 1, 2, 3, 4


def _fields(doctype: str, listed: str) -> list[str]:
	"""The fields of one screen that a `like` can ask about.

	`name` first and always — the id is what people paste, and a screen whose
	columns are eleven text fields would otherwise spend the whole budget before
	reaching it.

	The rest are the screen's own columns, filtered exactly as the in-screen
	search box filters them. Same constants, imported rather than restated: a
	field nobody may search on one screen is not searchable from the palette
	either, and two copies of that list is one copy that goes stale.
	"""
	try:
		meta = frappe.get_meta(doctype)
	except Exception:
		# A screen naming a doctype this site has not installed. The manifest
		# guards catch that; here it is one screen that answers nothing rather
		# than a palette that answers nothing.
		return []

	reach = ["name"]
	for fieldname in (listed or "").split(","):
		fieldname = fieldname.strip()
		if not fieldname or fieldname in reach:
			continue
		field = meta.get_field(fieldname)
		if not field or field.fieldtype in NEVER_SEARCHED:
			continue
		if "like" not in fieldtypes.operators_for(field.fieldtype):
			continue
		reach.append(fieldname)
		if len(reach) >= MAX_SEARCH_COLUMNS:
			break
	return reach


def _title_field(doctype: str, listed: list[str]) -> str:
	"""What to show as the hit's name.

	The doctype's own title field where it has one and the screen lists it —
	a hit reading `HR-EMP-00013` when the screen would have said `Bashir Haddad`
	is a result somebody has to open to identify.
	"""
	try:
		title = frappe.get_meta(doctype).title_field
	except Exception:
		return ""
	return title if title and title in listed else ""


def targets() -> list[dict]:
	"""Every screen this reader could open, as something to search.

	Deduplicated by doctype **within** a space and not across them, which is the
	honest answer to a doctype two spaces both list: a Sales Invoice really is
	reachable in OneBook and in RUA, they are different screens with different
	filters, and collapsing them would silently pick one. The same *record*
	turning up twice is handled where the results are ranked, not here.

	In manifest order, so the first screen a space declares over a doctype is
	the one its hits open in — which is the general list rather than a twin
	narrowed to one status.
	"""
	from oneapp.onespace import sync
	from oneapp.onespace.spaceview import navigable, visible

	found, seen = [], set()
	for space in visible(sync.state().get("spaces") or []):
		code = space.get("space_code") or ""
		for screen in navigable(space):
			doctype = (screen.get("document_type") or "").strip()
			if not doctype or (code, doctype) in seen:
				continue
			seen.add((code, doctype))
			found.append({
				"doctype": doctype,
				"space": code,
				"space_label": space.get("space_label") or code,
				"brand": space.get("brand") or "",
				"screen": screen.get("screen") or "",
				"label": screen.get("singular") or screen.get("label") or doctype,
				"icon": screen.get("icon") or "",
				"fields": screen.get("fields") or "",
				"filters": screen.get("filters") or "",
			})
			if len(found) >= SCREENS:
				return found
	return found


def _narrowing(target: dict) -> list:
	"""The screen's own filters, as `get_list` takes them.

	Through `mine.resolve`, so `@me` means the reader here exactly as it does
	when the screen itself is opened. A screen whose filters will not parse is
	searched unnarrowed rather than skipped — the permission check underneath
	is what keeps that safe, and the alternative is a screen that silently
	stops being findable.
	"""
	raw = target.get("filters")
	if not raw:
		return []
	try:
		declared = frappe.parse_json(raw) if isinstance(raw, str) else raw
	except Exception:
		return []
	if not isinstance(declared, dict):
		return []
	resolved = mine.resolve(declared)
	return [[fieldname, "=", value] if not isinstance(value, (list, tuple))
	        else [fieldname, value[0], value[1]]
	        for fieldname, value in resolved.items()]


def _where(text: str, title: str) -> int:
	"""How well one hit matches — `EXACT` down to `ELSEWHERE`."""
	low = (title or "").lower()
	if low == text:
		return EXACT
	if low.startswith(text):
		return STARTS
	if f" {text}" in f" {low}":
		return WORD
	return ANYWHERE if text in low else ELSEWHERE


def _hits(target: dict, text: str) -> list[dict]:
	"""One screen's answer, or nothing."""
	listed = _fields(target["doctype"], target["fields"])
	if not listed:
		return []

	title = _title_field(target["doctype"], listed)
	wanted = ["name", title] if title else ["name"]

	try:
		rows = frappe.get_list(
			target["doctype"],
			filters=_narrowing(target),
			or_filters=[[one, "like", f"%{text}%"] for one in listed],
			fields=wanted,
			order_by="modified desc",
			limit_page_length=MOST,
		)
	except frappe.PermissionError:
		# A screen whose grant and whose doctype permissions disagree. Absent
		# from the answer, which is what `navigable` would have done had it
		# known — never an error, because one bad screen must not take the
		# palette down with it.
		return []
	except Exception:
		frappe.clear_messages()
		return []

	found = []
	for seat, row in enumerate(rows):
		# Stripped, because a title field is not always plain text: OneTask's
		# subject is rich text, and a palette line reading
		# `<p>Chase the Halloway invoice</p>` is markup somebody has to read
		# past.
		shown = strip_html_tags(str(row.get(title) or row["name"])) if title else str(row["name"])
		shown = " ".join(shown.split()) or str(row["name"])
		found.append({
			"doctype": target["doctype"],
			"name": row["name"],
			"title": shown,
			# The id beside the title, and only where it is not the title
			# already — `TASK-0007 · TASK-0007` is a line that says one thing
			# twice.
			"id": row["name"] if shown != row["name"] else "",
			"space": target["space"],
			"space_label": target["space_label"],
			"brand": target["brand"],
			"screen": target["screen"],
			"label": target["label"],
			"icon": target["icon"],
			"rank": min(_where(text, shown), _where(text, row["name"])),
			# Where this hit sat in its own screen's answer. Read by the sort
			# below, which takes every screen's best hit before anybody's
			# second — a palette wants breadth before depth, and without it
			# one screen with five equally weak matches owns the visible list.
			"seat": seat,
		})
	return found


@frappe.whitelist(methods=["GET"])
def look(query: str = "", space: str = "") -> dict:
	"""What this person can find, from anywhere, for what they typed.

	`space` is where they were standing. It does not narrow anything — the
	whole point is that it does not — but it breaks the tie when one record is
	reachable from two spaces, and it lifts its own space's hits above equally
	good ones elsewhere. Somebody searching from inside OneBook means the
	invoice in OneBook.
	"""
	text = _search_text(query).strip().lower()
	if len(text) < SHORTEST:
		return {"query": text, "results": [], "looked": 0}

	found, looked = [], 0
	for target in targets():
		looked += 1
		found.extend(_hits(target, text))
		if len(found) >= TOTAL * 2:
			# Enough to rank well without walking the rest of the workspace.
			# Twice the ceiling rather than the ceiling, so the sort below is
			# still choosing between hits rather than reporting arrival order.
			break

	# One record, one line. Two spaces listing the same doctype each found it;
	# the one the reader is standing in wins, then manifest order.
	best: dict[tuple, dict] = {}
	for hit in found:
		key = (hit["doctype"], hit["name"])
		there = best.get(key)
		if there is None or (hit["space"] == space and there["space"] != space):
			best[key] = hit

	results = sorted(
		best.values(),
		key=lambda hit: (hit["rank"], hit["seat"],
		                 0 if hit["space"] == space else 1, hit["title"].lower()),
	)
	return {"query": text, "results": results[:TOTAL], "looked": looked}

File: frontend/src/modules/onespace/lib/shell/finding.js
/**
 * The one box, and the two halves of what it answers.
 *
 * **Where to go** needs no server at all. `api.session` already hands the
 * browser every space this person may open with the screens their seat allows,
 * so the rail, flattened, *is* the go-to list — filtered in memory, on the
 * keystroke, with nothing on the wire. That is why the box is useful the
 * instant it opens rather than after a round trip.
 *
 * **What to find** is `onespace/finding.py`, which fans a `like` out over
 * every screen the same person could open and hands back hits that already
 * know which space and screen they belong to.
 *
 * Both are ranked the same way and sorted into one list, because a person
 * typing `inv` means either and should not have to say which. Screens come
 * first at equal rank: there are a handful of them, they are what the word
 * usually meant, and a destination is cheaper to be wrong about than a record.
 */
import { computed, reactive } from 'vue'

import { session } from '@/modules/onespace/lib/shell/session'
import { workspace } from '@/shared/lib/workspace'
import { KIND, withAt } from '@/shared/lib/url/at'

/** Nothing is searched for less than this — the server's own floor. */
export const SHORTEST = 2

/** Screens offered before anybody types, and after. */
const SCREENS_SHOWN = 8

/** Where a match landed, best first — the same four the server ranks by. */
export const EXACT = 0
export const STARTS = 1
export const WORD = 2
export const ANYWHERE = 3
const NOWHERE = 9

/** How well one label matches, or `NOWHERE`. */
export function where(text, label) {
  const asked = String(text || '').trim().toLowerCase()
  const low = String(label || '').toLowerCase()
  if (!asked) return ANYWHERE
  if (low === asked) return EXACT
  if (low.startsWith(asked)) return STARTS
  if (` ${low}`.includes(` ${asked}`)) return WORD
  return low.includes(asked) ? ANYWHERE : NOWHERE
}

/**
 * Every screen this person could open, as one flat list.
 *
 * A component screen is in it and a screen with a doctype is in it, because
 * going somewhere does not care which: Configuration is a place in OneBook
 * exactly as Invoices is.
 */
export function screensOf(spaces) {
  const found = []
  for (const space of spaces || []) {
    for (const screen of space.screens || []) {
      found.push({
        kind: 'screen',
        key: `screen:${space.space_code}/${screen.screen}`,
        title: screen.label || screen.screen,
        space: space.space_code,
        spaceLabel: space.space_label || space.space_code,
        brand: space.brand || '',
        screen: screen.screen,
        icon: screen.icon || 'lucide-layout-grid',
      })
    }
  }
  return found
}

/**
 * The screens matching what was typed, best first.
 *
 * With nothing typed it is **where you already are**. The first eight screens
 * of a workspace with nine spaces in it are whichever two spaces sort first,
 * which is a list of places nobody asked about — and the box opens on it, so
 * it is the first thing anybody sees. The space you are standing in is the one
 * answer that is never arbitrary.
 */
export function screensMatching(screens, text, from = '') {
  const asked = String(text || '').trim()
  if (!asked) {
    const here = (screens || []).filter((one) => !from || one.space === from)
    return (here.length ? here : screens || []).slice(0, SCREENS_SHOWN)
  }
  const scored = (screens || [])
    .map((one) => ({
      ...one,
      rank: Math.min(where(asked, one.title), where(asked, one.spaceLabel)),
    }))
    .filter((one) => one.rank !== NOWHERE)
  scored.sort((a, b) => a.rank - b.rank || a.title.localeCompare(b.title))
  return scored.slice(0, SCREENS_SHOWN)
}

/** A record hit from the server, as a row this list can draw. */
export function recordRow(hit) {
  return {
    kind: 'record',
    key: `record:${hit.doctype}/${hit.name}`,
    title: hit.title,
    id: hit.id || '',
    label: hit.label || hit.doctype,
    space: hit.space,
    spaceLabel: hit.space_label || hit.space,
    brand: hit.brand || '',
    screen: hit.screen,
    icon: hit.icon || 'lucide-file-text',
    name: hit.name,
    rank: hit.rank ?? ANYWHERE,
  }
}

/**
 * Screens and records in one list.
 *
 * Sorted by how well each matched, and a screen ahead of a record at the same
 * rank. Not interleaved further than that: the server has already taken every
 * screen's best hit before anybody's second, so the order it sent is the order
 * that is kept inside a rank.
 */
export function merge(screens, records) {
  const rows = [
    ...screens.map((one, at) => ({ ...one, seat: at, first: 0 })),
    ...records.map((one, at) => ({ ...one, seat: at, first: 1 })),
  ]
  rows.sort((a, b) => a.rank - b.rank || a.first - b.first || a.seat - b.seat)
  return rows
}

/** Where one row goes. */
export function routeFor(row) {
  const query = { screen: row.screen }
  return {
    name: 'Screen',
    params: { spaceCode: row.space },
    query: row.kind === 'record' ? withAt(query, KIND.RECORD, row.name) : query,
  }
}

//: Open state, what was typed, and what came back. One object for the session,
//: because the box is the shell's and there is only ever one of it.
export const finder = reactive({
  open: false,
  typed: '',
  results: [],
  looking: false,
  failed: '',
  //: Where the box was opened from, which breaks a tie and narrows nothing.
  from: '',
})

export const screens = computed(() => screensOf(session.spaces))

export const shown = computed(() =>
  merge(
    screensMatching(screens.value, finder.typed, finder.from),
    finder.results.map(recordRow),
  ),
)

export function openFinder(from = '') {
  finder.from = from || ''
  finder.open = true
}

export function closeFinder() {
  finder.open = false
  finder.typed = ''
  finder.results = []
  finder.failed = ''
  finder.looking = false
}

//: Only the newest answer is kept. A search is debounced but not serialised,
//: so a slow request for `inv` can land after a fast one for `invoice` and
//: repaint the list with the older word's hits.
let asked = 0

export async function look(text) {
  const query = String(text || '').trim()
  if (query.length < SHORTEST) {
    finder.results = []
    finder.failed = ''
    finder.looking = false
    return
  }
  const mine = ++asked
  finder.looking = true
  try {
    const answer = await workspace.find(query, finder.from)
    if (mine !== asked) return
    finder.results = answer?.results || []
    finder.failed = ''
  } catch (error) {
    if (mine !== asked) return
    finder.results = []
    finder.failed = String(error?.message || error || '')
  } finally {
    if (mine === asked) finder.looking = false
  }
}
